tools/jnimarshalmethod-gen/App.cs

```diff
@@ -133,9 +244,192 @@ static void AddRegisterNativeMembers (TypeBuilder dt, ParameterExpression target
 			var lambda  = Expression.Lambda<Action<JniNativeMethodRegistrationArguments>> (body, new[]{ args });
 
 			var rb = dt.DefineMethod ("__RegisterNativeMembers",
-					MethodAttributes.Public | MethodAttributes.Static);
+					System.Reflection.MethodAttributes.Public | System.Reflection.MethodAttributes.Static);
 			rb.SetCustomAttribute (new CustomAttributeBuilder (typeof (JniAddNativeMethodRegistrationAttribute).GetConstructor (Type.EmptyTypes), new object[0]));
 			lambda.CompileToMethod (rb);
 		}
+
+		public static void Error (string message)
+		{
+			Console.Error.WriteLine ($"{Terminal.FgColor (Color.Red, true)}Error: {Name}: {message}{Terminal.ResetColor (true)}");
+		}
+
+		public static void Warning (string message)
+		{
+			Console.Error.WriteLine ($"{Terminal.FgColor (Color.Yellow, true)}Warning: {Name}: {message}{Terminal.ResetColor (true)}");
+		}
+	}
+
+	static class Extensions
+	{
+		public static string GetCecilName (this Type type)
+		{
+			return type.FullName.Replace ('+', '/');
+		}
+
+		static bool CompareTypes (Type reflectionType, TypeReference cecilType)
+		{
+			return cecilType.ToString () == reflectionType.GetCecilName ();
+		}
+
+		static bool MethodsAreEqual (MethodInfo methodInfo, MethodDefinition methodDefinition)
+		{
+			if (methodInfo.Name != methodDefinition.Name)
+				return false;
+
+			if (methodInfo.GetParameters ().Length != methodDefinition.Parameters.Count)
```

Comment:
This should probably check `methodInfo.HasParameters == methodDefinition.HasParameters` *first*.

Reply:
MethodInfo doesn't have `HasParameters`, I changed it anyway to at least use `methodDefinition.HasParameters` and added null checking.